### PR TITLE
Assertj

### PR DESCRIPTION
**Pull request by andricDu**

### CODE REVIEW

File: _posts/dusan_andric/2019-08-19-assertJ.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+Yes.
+
+<figure>
+    <img src="{{site.urlimg}}dusan_andric/assertj/assertJ_pr.png" />
+    <figcaption>The first attempt at fixing compilation times.</figcaption>
+</figure>
+
+## But why?
+
+Being completely honest, we do not have a have the definitive answer to what was causing this, however, I do have a hunch.
```
Comment (ciaranschutte) [inline _posts/dusan_andric/2019-08-19-assertJ.md]: "we do not have a have the" => "we do not have the def..."

File: _posts/dusan_andric/2019-08-19-assertJ.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+We know that this became a big problem for us when we switched to Java 11. We also know that it didn't matter if we were using Oracle JDK or OpenJDK as they are within margin of error of one another. One of our team members, [Kevin](http://softeng.oicr.on.ca/blog/category/kevin_hartmann), discovered previously reported issues in JDK8 about a regression in compliation times, as well as him discovering other various issues and creating bug reports for `javac`.
```
Comment (ciaranschutte) [inline _posts/dusan_andric/2019-08-19-assertJ.md]: "within margin" => "within a margin.."

File: _posts/dusan_andric/2019-08-19-assertJ.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+A big reason for us using AssertJ, in addition to handling exception better than JUnit4, is the fluent API it had made it a lot more ergonomic to use for developers. Code looked better, was easier to understand, and it made reuse easier.
+
+If you are unfamiliar with the API of AssertJ, I have an example here:
+~~~java
+    assertThat(myList).hasSize(n).containsAnyOf(matchObj)
+~~~
+
+You can imagine that `assertThat` not only being a heavily overloaded method and making use of generics, also allows you to branch out to many different similarly generic, heavily overloaded methods that are context sensitive on the return type of the previous callable.
+
+Another fact to consider is that `javac` is not sitting idle during this long compilation time, its going full tilt on my machine maxing out the cores it is running on.
```
Comment (ciaranschutte) [inline _posts/dusan_andric/2019-08-19-assertJ.md]: its going
it's going

File: _posts/dusan_andric/2019-08-19-assertJ.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+My hypothesis is that at some stage in `javac`, either type inference, or during the type erasure where it is bounding the generic types, that the root of the problem lies. It must be some computation of high complexity, maybe exponential based on the hight of the type hierarchy for AssertJ.
```
Comment (ciaranschutte) [inline _posts/dusan_andric/2019-08-19-assertJ.md]: hight  => height?

File: _posts/dusan_andric/2019-08-19-assertJ.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    <figcaption>The first attempt at fixing compilation times.</figcaption>
+</figure>
+
+## But why?
+
+Being completely honest, we do not have a have the definitive answer to what was causing this, however, I do have a hunch.
+
+We know that this became a big problem for us when we switched to Java 11. We also know that it didn't matter if we were using Oracle JDK or OpenJDK as they are within margin of error of one another. One of our team members, [Kevin](http://softeng.oicr.on.ca/blog/category/kevin_hartmann), discovered previously reported issues in JDK8 about a regression in compliation times, as well as him discovering other various issues and creating bug reports for `javac`.
+
+Given this lack of polish with `javac` in Java 11 and knowing there were previous type inference issues that were probably exaserbated with the addition of the `var` keyword in Java 10, I figured that type inference had to be related to the issue, or at the very least the type system itself.
```
Comment (samrichca) [inline _posts/dusan_andric/2019-08-19-assertJ.md]:
```suggestion
Given this lack of polish with `javac` in Java 11 and knowing there were previous type inference issues that were probably exacerbated with the addition of the `var` keyword in Java 10, I figured that type inference had to be related to the issue, or at the very least the type system itself.
```

File: _posts/dusan_andric/2019-08-19-assertJ.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+A big reason for us using AssertJ, in addition to handling exception better than JUnit4, is the fluent API it had made it a lot more ergonomic to use for developers. Code looked better, was easier to understand, and it made reuse easier.
```
Comment (samrichca) [inline _posts/dusan_andric/2019-08-19-assertJ.md]:
```suggestion
A big reason for us using AssertJ, in addition to handling exception better than JUnit4, is the fluent API that had made it a lot more ergonomic to use for developers. Code looked better, was easier to understand, and it made reuse easier.
```

File: _posts/dusan_andric/2019-08-19-assertJ.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+Another fact to consider is that `javac` is not sitting idle during this long compilation time, its going full tilt on my machine maxing out the cores it is running on.
+
+My hypothesis is that at some stage in `javac`, either type inference, or during the type erasure where it is bounding the generic types, that the root of the problem lies. It must be some computation of high complexity, maybe exponential based on the hight of the type hierarchy for AssertJ.
+
+## Do I have this problem too?
+
+One of the symptoms we noticed is that opening a project and editing a file in IntelliJ Idea, the IDE would lint/inspect the file at a snails pace. For example if I removed an important import and hit `ctrl+s`, for several seconds the IDE would slowly make its way through the file and I could see the red ticks on the scroll bar crawl it's way down.
+
+If this truly is related to complex type hierarchies and generics I would love it other people reached out and gave us examples outside of AssertJ if for nothing more than building an understanding of what the root of the problem really is. I'd be just as happy if you were to debunk my hypothesis.
```
Comment (ciaranschutte) [inline _posts/dusan_andric/2019-08-19-assertJ.md]: i would love it other people
i would love it if other people

File: _posts/dusan_andric/2019-08-19-assertJ.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+You can imagine that `assertThat` not only being a heavily overloaded method and making use of generics, also allows you to branch out to many different similarly generic, heavily overloaded methods that are context sensitive on the return type of the previous callable.
+
+Another fact to consider is that `javac` is not sitting idle during this long compilation time, its going full tilt on my machine maxing out the cores it is running on.
+
+My hypothesis is that at some stage in `javac`, either type inference, or during the type erasure where it is bounding the generic types, that the root of the problem lies. It must be some computation of high complexity, maybe exponential based on the hight of the type hierarchy for AssertJ.
+
+## Do I have this problem too?
+
+One of the symptoms we noticed is that opening a project and editing a file in IntelliJ Idea, the IDE would lint/inspect the file at a snails pace. For example if I removed an important import and hit `ctrl+s`, for several seconds the IDE would slowly make its way through the file and I could see the red ticks on the scroll bar crawl it's way down.
```
Comment (hlminh2000) [inline _posts/dusan_andric/2019-08-19-assertJ.md]: `at a snails pace` => `at a snail's pace`

File: _posts/dusan_andric/2019-08-19-assertJ.md (hunk 4 above)
Comment (samrichca) [inline _posts/dusan_andric/2019-08-19-assertJ.md]: `hight of the type hierarchy`

what's hight?

File: _posts/dusan_andric/2019-08-19-assertJ.md (hunk 8 above)
Comment (samrichca) [inline _posts/dusan_andric/2019-08-19-assertJ.md]:
```suggestion
One of the symptoms we noticed is that when opening a project and editing a file in IntelliJ Idea, the IDE would lint/inspect the file at a snail's pace. For example if I removed an important import and hit `ctrl+s`, for several seconds the IDE would slowly make its way through the file and I could see the red ticks on the scroll bar crawl its way down.
```

File: _posts/dusan_andric/2019-08-19-assertJ.md (hunk 7 above)
Comment (hlminh2000) [inline _posts/dusan_andric/2019-08-19-assertJ.md]: `I would love it other people reached out` -> `I would love it IF other people reached out`

Comment (samrichca) [inline _posts/dusan_andric/2019-08-19-assertJ.md]:
```suggestion
If this truly is related to complex type hierarchies and generics I would love it if other people reached out and gave us examples outside of AssertJ, if for nothing more than building an understanding of what the root of the problem really is. I'd be just as happy if you were to debunk my hypothesis.
```

**Review by KevinFHartmann — APPROVED** [review]

**Pull request merged**